cmd/kubeadm/app/util/config/initconfiguration.go

```diff
@@ -128,14 +128,23 @@ func SetAPIEndpointDynamicDefaults(cfg *kubeadmapi.APIEndpoint) error {
 
 // SetClusterDynamicDefaults checks and sets values for the ClusterConfiguration object
 func SetClusterDynamicDefaults(cfg *kubeadmapi.ClusterConfiguration, advertiseAddress string, bindPort int32) error {
+	kubeProxyUserBindAddress := ""
+	if cfg.ComponentConfigs.KubeProxy != nil && cfg.ComponentConfigs.KubeProxy.BindAddress != "" {
+		kubeProxyUserBindAddress = cfg.ComponentConfigs.KubeProxy.BindAddress
+	}
+
 	// Default all the embedded ComponentConfig structs
+	// This will set 'cfg.ComponentConfigs.KubeProxy.BindAddress' as '0.0.0.0' if it is unset
 	componentconfigs.Known.Default(cfg)
 
-	ip := net.ParseIP(advertiseAddress)
-	if ip.To4() != nil {
-		cfg.ComponentConfigs.KubeProxy.BindAddress = kubeadmapiv1beta2.DefaultProxyBindAddressv4
-	} else {
-		cfg.ComponentConfigs.KubeProxy.BindAddress = kubeadmapiv1beta2.DefaultProxyBindAddressv6
+	// Since kube-proxy is a daemonset, so the bind address only has two options: either 0.0.0.0 or ::
+	if kubeProxyUserBindAddress == "" || (kubeProxyUserBindAddress != kubeadmapiv1beta2.DefaultProxyBindAddressv4 && kubeProxyUserBindAddress != kubeadmapiv1beta2.DefaultProxyBindAddressv6) {
```

Comment:
IMO we can also use DefaultProxyBindAddressv6 here wdyt?

Reply:
bind address was already exposed to users, so it is better to allow customizing it.

For some special users, they may only want to bind to IPv4 address.

Reply:
> bind address was already exposed to users, so it is better to allow customizing it.

yes.
this will overwrite any custom user value, which is not correct even if the DS address should point to the default interface:

why can't we just use this?
```
if kubeProxyUserBindAddress == "" {
	ip := net.ParseIP(advertiseAddress)
	if ip.To4() != nil {
		cfg.ComponentConfigs.KubeProxy.BindAddress = kubeadmapiv1beta2.DefaultProxyBindAddressv4
	} else {
		cfg.ComponentConfigs.KubeProxy.BindAddress = kubeadmapiv1beta2.DefaultProxyBindAddressv6
	}
}
```

also remove this comment:
```
// Since kube-proxy is a daemonset, so the bind address only has two options: either 0.0.0.0 or ::
```